runtime/compiler/trj9/runtime/ClassUnloadAssumption.cpp

```diff
@@ -351,6 +354,149 @@ TR_RuntimeAssumptionTable::detachFromRAT(OMR::RuntimeAssumption *assumption)
+                  }
+               }
+            TR_ASSERT( hashTable->_markedforDetachCount[i]==0, "RAT detach count should be 0!" );
+            }
+         _detachPending[kind] = false;
+         }
+      }
+   assumptionTableMutex->exit();
+   TR_ASSERT( marked == reclaimed, "Did not reclaime all the marked assumptions!" );
+   marked=0;
+   }
+
+/**
+ * Mark and detach all of the assumptions in the metadata's circular linked list
+ * @param md The metadata for which to mark & detach assumptions from.
+ * @param reclaimPrePrologueAssumptions Are all assumptions free to be removed (ClassUnloading vs. recompiled)
+ * This only detaches from the metadata's linked list, not a detech from the RAT,
+ * to remove from the RAT you need to call reclaimMarkedAssumptionsFromRAT()
+ */
+void TR_RuntimeAssumptionTable::markAssumptionsAndDetach(void * md, bool reclaimPrePrologueAssumptions)
+   {
+   J9JITExceptionTable *metaData = (J9JITExceptionTable*) md;
+   OMR::RuntimeAssumption *sentry = (OMR::RuntimeAssumption*)(metaData->runtimeAssumptionList);
+   OMR::RuntimeAssumption *cursor, *next;
+
+   assumptionTableMutex->enter();
+   if (sentry)
+      {
+      int32_t numAssumptionsNotReclaimed = 0; // RAS
+      TR_ASSERT(sentry->getAssumptionKind() == RuntimeAssumptionSentinel, "First assumption must be the sentinel\n");
+      OMR::RuntimeAssumption *notReclaimedList = sentry;
+      for ( cursor=sentry->getNextAssumptionForSameJittedBody(); cursor != sentry; cursor=next)
+         {
+         next = cursor->getNextAssumptionForSameJittedBody();
+         if (!metaData || cursor->isAssumingMethod(metaData, reclaimPrePrologueAssumptions))
+            {
+            markForDetachFromRAT(cursor); // Mark for deletion
+            marked++;
+            }
+         else // This could be an assumption to the persistentBodyInfo which is not reclaimed
+            {
+            cursor->setNextAssumptionForSameJittedBody(notReclaimedList);
+            notReclaimedList = cursor;
+            numAssumptionsNotReclaimed++; // RAS
+            }
+         }
+
+      if (notReclaimedList != sentry) // some entries were not reclaimed
+         {
+         // Must attach the notReclaimedList to the sentinel
+         sentry->setNextAssumptionForSameJittedBody(notReclaimedList);
+
+#ifdef PROD_WITH_ASSUMES
+         if (numAssumptionsNotReclaimed > 1) 
+            {
+            cursor = sentry->getNextAssumptionForSameJittedBody();
+            while (cursor != sentry)
+               {
+               TR_RuntimeAssumptionKind kind = cursor->getAssumptionKind();
+               if (kind != RuntimeAssumptionOnClassRedefinitionPIC && 
+                   kind != RuntimeAssumptionOnClassRedefinitionUPIC && 
+                   kind != RuntimeAssumptionOnClassRedefinitionNOP)
+                  {
+                  fprintf(stderr, "%d assumptions were left after metadata %p assumption reclaiming\n", numAssumptionsNotReclaimed, metaData);
```

Comment:
fprintf to stderr is not very elegant for debug code - this should either use the tracing framework or the information integrated into the assert message.

Reply:
This part of the code is a direct copy of the existing code in TR_RuntimeAssumptionTable::reclaimAssumptions() which could be made redundant in the future if we decide that this mark@sweep methods is appropriate for the compile abort cases (and I believe it is).  I left this code as is so that it's behaviour (and output) is identical to what it was before.

Reply:
@mpirvu is there a better way that fprintf that should be used here? I'm just not a fan of hardcoded prints to stderr because they defeat so much of the infrastructure around gathering diagnostics and output. If this is the 'right' way then I'll relent.

Reply:
I am in favor of removing the entire section because we no longer build/test PROD_WITH_ASSUMES

Reply:
I'd prefer to keep it with both PROD_WITH_ASSUMES and DEBUG since it is a useful validation that the code is running right... Some people do manually build prod with assumes still I believe.

Reply:
ok, then the fprintf could become a TR_ASSERT

Reply:
I decided to remove the entire #define section and just add an assert into the main loop. Seems easier and cleaner.